### DISCUSSION TITLE
What are the DomainEvents in the TodoItem

### DISCUSSION THREAD

**Discussion by TCrosmanDSI** [jasontaylordev/CleanArchitecture#426] [General]
I'm assuming it's for some sort of record keeping? I've looked through the code and can't find where the DomainEvents for the ToDoItem do anything.

Can someone explain what these are for to me?

**Comment by ibrahimuludag** [comment]
Hi,

An event is something that has happened in the past. A domain event is, something that happened in the domain that you want other parts of the same domain (in-process) to be aware of. The notified parts usually react somehow to the events.

An important benefit of domain events is that side effects can be expressed explicitly.

For example, if you're just using Entity Framework and there has to be a reaction to some event, you would probably code whatever you need close to what triggers the event. So the rule gets coupled, implicitly, to the code, and you have to look into the code to, hopefully, realize the rule is implemented there.

On the other hand, using domain events makes the concept explicit, because there is a DomainEvent and at least one DomainEventHandler involved.

You can find more information here:
- https://martinfowler.com/eaaDev/DomainEvent.html
- https://docs.microsoft.com/en-us/dotnet/architecture/microservices/microservice-ddd-cqrs-patterns/domain-events-design-implementation

**Reply by TCrosmanDSI** [reply]
Thank you for the explanation.

If possible...could you explain to me what happens with the domain event in the todoitem example in this project?

I see it get's added in the create command and gets set in the setter of the "done" property, and I see that the DbContext calls an event dispatcher, but I can't find the code that shows what happens with the dispatched events of the todoitem or where any other entities are subscribed to these events.

**Reply by TCrosmanDSI** [reply]
Ok I think I understand now.

Tell me if this is a correct presumption.

The point of domain events is to fire events on entity framework "Save"?  They are just regular events, but we don't fire them from the entities themselves?

**Reply by Eagleleader** [reply]
In this project, they attached domain events to entities (if you need one). In the save method, they check for each entity if there is any domain event attached to it and they publish it.
They then, retrieve it into the Handler to log some information about it but of course, you can do whatever you want.

**Reply by AlexAlexGoTO** [reply]
Thanks for explanation
